### ISSUE TITLE
Progress Events?

### ISSUE THREAD

**Issue by EddieOne** [theophilusx/ssh2-sftp-client#200]
Hello, Is there any way to get a progress event for uploadDir. I'm testing that on 5.x and it seems to work well except the user doesn't know what's happening. :)

**Comment by theophilusx** [comment]
That seems reasonable. I should be able to add/emit an upload progress
event which client code could add a handler for to get progress
data. Will let you know when I've done it and updated the branch.


-- 
Tim Cross

**Comment by theophilusx** [comment]
I have committed some changes which add an 'upload' event to uploadDir and a 'download' event to downloadDir. Both events call listeners with a single argument, which is an object that has properties for source and destination after each file has bene successfully uploaded/downloaded. There are examples in the README and two example scripts in the example directory of the repository. 

I thought about trying to do something which provided details of percentage of download/upload completed, but it was too much of a performance hit - especially for donwloadDir. The main problem is that because the methods allow upload/download of sub directories, they work in a recursive nature. To calculate total numbers, you need to make multiple list() requests. Having an event which just rports once a file has been downloaded seems like a reasonable compromise

**Comment by EddieOne** [comment]
Alright, thank you. I have tested the code change and it works. It does send a filename back. So if I count the files using Node, I should be able to display a progress bar.

I was thinking how to send this info to the browser and I saw one module called Formidable that is using xhr.upload and some magic to get progress events to browsers. Maybe a persistent socket would simplify the browser interaction. I will have to keep testing. For now, we can close this task.
